tests/ui/type_complexity_issue_1013.rs

```diff
@@ -0,0 +1,23 @@
+#![warn(clippy::type_complexity)]
+use std::iter::{Filter, Map};
+use std::vec::IntoIter;
+
+struct S;
+
+impl IntoIterator for S {
+    type Item = i32;
+    // Should not warn since there is no way to simplify this
+    type IntoIter = Filter<Map<IntoIter<i32>, fn(i32) -> i32>, fn(&i32) -> bool>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        fn m(a: i32) -> i32 {
+            a
+        }
+        fn p(_: &i32) -> bool {
+            true
+        }
+        vec![1i32, 2, 3].into_iter().map(m as fn(_) -> _).filter(p)
+    }
```

Comment:
Is this "complicated" implementation necessary to test the lint? I think you could also implement the test without using `std` traits, but rather write a "dummy" trait yourself, with just one assoc type in it. 

---

One a related note: Please put this test into the test file that already exists for this lint.

Reply:
I've reused the dummy trait that was already present in the test 👍🏻